core/src/main/scala/chisel3/connectable/Connectable.scala

```diff
@@ -54,6 +54,9 @@ final class Connectable[+T <: Data] private (
     this.copy(waived = waivedMembers.toSet) // not appending waived because we are collecting all members
   }
 
+  /** Waive all members of base and static cast to a new type */
+  def waiveAllAs[S <: Data]: Connectable[S] = waiveAll.asInstanceOf[Connectable[S]]
```

Comment:
I think you may want some proof that `T <:< S` to let people call these functions. What happens if I do something like:
```scala
val foo = IO(Decoupled(UInt(8.W)))
foo.waiveAllAs[UInt]
```

Reply:
You'd get a runtime error, but that's the same as if you used `.asInstanceOf[UInt]` right?

Reply:
Can you add a test showing the behavior? I'm not convinced you get a runtime error (because the cast is about a type parameter and those are erased at runtime), you might just get silently dropped connections

Reply:
Oh, you are right. There is no error. Ok, I'll add the super type constraint implicit.

Reply:
Actually, in my experience, having the constraint actually isn't a good idea. For example, you may want to connected a Decoupled to a Valid. Since DecoupledIO is not a subtype of Valid, you still may want to do:
```
val foo: DecoupledIO[UInt] = ???
val bar: Valid[UInt] = ???
foo.waiveAs[Valid[UInt]](_.ready) :<>= bar
```

Reply:
You could try our shiny `ChiselSubtypeOf` https://github.com/chipsalliance/chisel3/blob/b90a803849724662c115e8bbb537587bdcec9f32/core/src/main/scala/chisel3/experimental/dataview/ChiselSubtypeOf.scala#L28

🙂